zarr/convenience.py

```diff
@@ -348,3 +350,392 @@ def load(store):
+
+
+def _copy(log, source, dest, name, root, shallow, without_attrs, **create_kws):
+
+    # are we copying to/from h5py?
+    source_h5py = source.__module__.startswith('h5py.')
+    dest_h5py = dest.__module__.startswith('h5py.')
+
+    # determine name to copy to
+    if name is None:
+        name = source.name.split('/')[-1]
+        if not name:
+            raise TypeError('source has no name, please provide the `name` '
+                            'parameter to indicate a name to copy to')
+
+    if hasattr(source, 'shape'):
+        # copy a dataset/array
+
+        # setup creation keyword arguments
+        kws = create_kws.copy()
+
+        # setup chunks option, preserve by default
+        kws.setdefault('chunks', source.chunks)
+
+        # setup compression options
+        if source_h5py:
+            if dest_h5py:
+                # h5py -> h5py; preserve compression options by default
+                kws.setdefault('compression', source.compression)
+                kws.setdefault('compression_opts', source.compression_opts)
+                kws.setdefault('shuffle', source.shuffle)
+            else:
+                # h5py -> zarr; use zarr default compression options
+                pass
+        else:
+            if dest_h5py:
+                # zarr -> h5py; use some vaguely sensible defaults
+                kws.setdefault('compression', 'gzip')
+                kws.setdefault('compression_opts', 1)
+                kws.setdefault('shuffle', True)
+            else:
+                # zarr -> zarr; preserve compression options by default
+                kws.setdefault('compressor', source.compressor)
+
+        # create new dataset in destination
+        ds = dest.create_dataset(name, shape=source.shape, dtype=source.dtype, **kws)
+
+        # copy data - N.B., if dest is h5py this will load all data into memory
+        log('{} -> {}'.format(source.name, ds.name))
+        ds[:] = source
+
+        # copy attributes
+        if not without_attrs:
+            ds.attrs.update(source.attrs)
+
+    elif root or not shallow:
+        # copy a group
+
+        # creat new group in destination
+        grp = dest.create_group(name)
+        log('{} -> {}'.format(source.name, grp.name))
+
+        # copy attributes
+        if not without_attrs:
+            grp.attrs.update(source.attrs)
+
+        # recurse
+        for k in source.keys():
+            _copy(log, source[k], grp, name=k, root=False, shallow=shallow,
+                  without_attrs=without_attrs, **create_kws)
+
+
+def tree(grp, expand=False, level=None):
+    """Provide a ``print``-able display of the hierarchy. This function is provided
+    mainly as a convenience for obtaining a tree view of an h5py group - zarr groups
+    have a ``.tree()`` method.
+
+    Parameters
+    ----------
+    grp : Group
+        Zarr or h5py group.
+    expand : bool, optional
+        Only relevant for HTML representation. If True, tree will be fully expanded.
```

Comment:
Sorry this is a bit off topic and I may have forgotten the answer, do we warn if `expand` is set to `True` for the text representation? If not, we might want to look into it.

Reply:
Revisiting this, it doesn't look like this would be easy to do as we don't know what representation will be used when `expand` is set.